ic-cdk-executor/src/machinery.rs

```diff
@@ -0,0 +1,474 @@
+    enter_current_method(guard, |_| {
+        let res = f();
+        poll_all();
+        res
+    })
+}
+
+/// Execute an inter-canister call callback in the context of the method that made it.
+pub fn in_callback_executor_context_for<R>(
+    method_handle: MethodHandle,
+    f: impl FnOnce() -> R,
+) -> R {
+    setup_panic_hook();
+    enter_current_method(method_handle, |_| {
+        let res = f();
+        poll_all();
+        res
+    })
+}
+
+/// Enters a panic recovery context for calling [`cancel_all_tasks_attached_to_current_method`] in.
+pub fn in_trap_recovery_context_for<R>(method: MethodHandle, f: impl FnOnce() -> R) -> R {
+    setup_panic_hook();
+    enter_current_method(method, |_| {
+        RECOVERING.set(true);
+        let res = f();
+        RECOVERING.set(false);
+        res
+    })
+}
+
+/// Cancels all tasks made with [`spawn_protected`] attached to the current method.
+pub fn cancel_all_tasks_attached_to_current_method() {
+    let Some(method_id) = CURRENT_METHOD.get() else {
+        panic!("`cancel_all_tasks_attached_to_current_method` can only be called within a method context");
+    };
+    cancel_all_tasks_attached_to_method(method_id);
+}
+
+/// Cancels all tasks made with [`spawn_protected`] attached to the given method.
+fn cancel_all_tasks_attached_to_method(method_id: MethodId) {
+    let to_cancel = METHODS.with_borrow_mut(|methods| {
+        let Some(method) = methods.get_mut(method_id) else {
+            panic!("internal error: method context deleted while in use (cancel_all_tasks_attached_to_method)");
+        };
+        take(&mut method.tasks)
+    });
+    let _tasks = TASKS.with(|tasks| {
+        let Ok(mut tasks) = tasks.try_borrow_mut() else {
+            panic!(
+                "`cancel_all_tasks_attached_to_current_method` cannot be called from an async task"
+            );
+        };
+        let mut canceled = Vec::with_capacity(to_cancel.len());
+        for task_id in to_cancel {
+            canceled.push(tasks.remove(task_id));
+        }
+        canceled
+    });
+    drop(_tasks); // always run task destructors outside of a refcell borrow
+}
+
+/// Cancels a specific task. Use this instead of `remove` for guaranteed drop order.
+pub(crate) fn cancel_task(task_id: TaskId) {
+    let _task = TASKS.with_borrow_mut(|tasks| tasks.remove(task_id));
+    drop(_task); // always run task destructors outside of a refcell borrow
+}
+
+/// Returns true if tasks are being canceled due to a trap or panic.
+pub fn is_recovering_from_trap() -> bool {
+    RECOVERING.get()
+}
+
+/// Produces a handle to the current method context.
+///
+/// The method is active as long as the handle is alive.
+pub fn extend_current_method_context() -> MethodHandle {
+    setup_panic_hook();
+    let Some(method_id) = CURRENT_METHOD.get() else {
+        panic!("`extend_method_context` can only be called within a tracking executor context");
+    };
+    MethodHandle::for_method(method_id)
+}
```

Comment:
This is a bit tricky to use correctly because dropping the last handle doesn't actually cause the methods to be canceled, right? That only happens if we complete `enter_context_for` with 0 handles. So if we really expect third parties to use this we should probably say more about how to use it properly.